crates/subspace-core-primitives/src/objects.rs

```diff
@@ -129,50 +129,55 @@ pub struct PieceObjectMapping {
 #[derive(Debug, Copy, Clone, PartialEq, Eq, Ord, PartialOrd, Hash, Encode, Decode, TypeInfo)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
-pub enum GlobalObject {
-    /// V0 of object mapping data structure
-    #[codec(index = 0)]
-    #[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
-    V0 {
-        /// Piece index where object is contained (at least its beginning, might not fit fully).
-        /// The index and offset must be first, so that the Ord derives are valid.
-        piece_index: PieceIndex,
-        /// Offset of the object in that piece
-        offset: u32,
-        /// Object hash
-        hash: Blake3Hash,
-    },
+pub struct GlobalObject {
+    /// Object hash.
+    /// We order by hash, so object hash lookups can be performed efficiently.
+    hash: Blake3Hash,
+    /// Piece index where object is contained (at least its beginning, might not fit fully)
+    piece_index: PieceIndex,
+    /// Offset of the object in that piece
+    offset: u32,
 }
 
 impl GlobalObject {
+    /// Returns a newly created GlobalObject from its fields.
+    pub fn new(hash: Blake3Hash, piece_index: PieceIndex, offset: u32) -> Self {
+        Self {
+            hash,
+            piece_index,
+            offset,
+        }
+    }
+
+    /// Object hash
+    pub fn hash(&self) -> Blake3Hash {
+        self.hash
+    }
+
     /// Piece index where object is contained (at least its beginning, might not fit fully)
     pub fn piece_index(&self) -> PieceIndex {
-        match self {
-            Self::V0 { piece_index, .. } => *piece_index,
-        }
+        self.piece_index
     }
 
     /// Offset of the object
     pub fn offset(&self) -> u32 {
-        match self {
-            Self::V0 { offset, .. } => *offset,
-        }
-    }
-
-    /// Object hash
-    pub fn hash(&self) -> Blake3Hash {
-        match self {
-            Self::V0 { hash, .. } => *hash,
-        }
+        self.offset
     }
 }
 
 /// Mapping of objects stored in the history of the blockchain
-#[derive(Debug, Default, Clone, PartialEq, Eq, Ord, PartialOrd, Hash, Encode, Decode, TypeInfo)]
+#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd, Hash, Encode, Decode, TypeInfo)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
-pub struct GlobalObjectMapping {
-    /// Objects stored in the history of the blockchain.
-    /// Mappings are ordered by the piece index and offset of the first GlobalObject in objects.
-    pub objects: Vec<GlobalObject>,
+#[cfg_attr(feature = "serde", serde(rename_all_fields = "camelCase"))]
+pub enum GlobalObjectMapping {
+    /// V0 of object mapping data structure.
+    /// The enum is outside the vector to save memory.
+    /// (We won't store different codecs in the same vector.)
```

Comment:
I'm not sure this is really a necessary comment. Also what you call "codec" is just an enum variant. `#[codec(index = 0)]` is actually the default here, we only specify it explicitly for long-term stability, see [docs](https://docs.rs/parity-scale-codec/latest/parity_scale_codec/#derive-attributes). No codec is added or removed here.